worker/runtime/cni_network.go

```diff
@@ -220,6 +250,20 @@ func (n cniNetwork) SetupMounts(handle string) ([]specs.Mount, error) {
 	}, nil
 }
 
+func (n cniNetwork) SetupRestrictedNetworks() error {
+	const tableName = "filter"
+	err := n.ipt.CreateChainOrFlushIfExists(tableName, ipTablesAdminChainName)
+
+	// Create REJECT rules in empty admin chain
```

Comment:
I remember when exploring the iptables rules in CF as part of the networking onboarding week, we encountered a rule that looks like:

```
-A w--default -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
```

https://github.com/cloudfoundry/guardian/blob/b9201b81a0a5adfe9cafec2cb954e676a28e567b/kawasaki/iptables/global_chains.go#L294

I believe this was an optimization to prevent checking the destination of every packet once the connection is already established (we only check on the initial handshake packets).

We should probably do the same

Reply:
Interesting point. I believe you're suggesting adding an "accept-all" rule for RELATED and ESTABLISHED connections before the REJECT rule. So the output using `CONCOURSE_GARDEN_DENY_NETWORK=152.34.4.5` would be

```
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
REJECT     all  --  anywhere             152.34.4.5      reject-with icmp-port-unreachable
```

I'm not aware of the exact mechanics of the kernel implementation of the connection tracking module. But from my understanding of `iptables`, for the first rule to work each packet, even when RELATED or ESTABLISHED, will be checked against the rule to ensure conformance. This results in an operation that likely involves checking the connection state.

Without this rule, each packet would be checked regardless of state against the `REJECT` rule to see if destination matches.

In that crude sense the performance should be roughly the same with two rules or one rule, as both cases involve a rule checking operation. If the connection status is NEW there would actually be an extra operation for the two rule case as two rules would have to be checked against.

Again, I say this without understanding the kernel [implementation](https://github.com/torvalds/linux/blob/f85c1598ddfe83f61d0656bd1d2025fa3b148b99/net/netfilter/nf_conntrack_core.c), and from not seeing performance benefits mentioned in the [docs](https://linux.die.net/man/8/iptables).

Reply:
I guess the idea would be that if there were 10 deny network ranges, we'd need to perform 10 checks for every packet that isn't denied, even though we could skip those checks for the vast majority of those packets.

e.g. if we're uploading a 1GB file, assuming an MTU of 1500 bytes/packet per https://stackoverflow.com/a/2614188/9994448, that means we're sending out few packets (maybe just 1?) to establish a connection, and ~600,000 packets with data. With the "accept all RELATED and ESTABLISHED" rule, we're doing ~600,000 checks - without it, we're doing ~6,000,000.

That said, it's probably a good idea to benchmark it in any case, or maybe talk to some Guardian folks to see why they went with that approach.

Reply:
Great point! I was about to reply with a similar realization about checks being made until a match happens. Even with no deny network ranges, and 10 running containers, a packet from an ESTABLISHED/RELATED connection in the 10th container (e.g. outbound from 10.80.0.10) would go through 20 rules before it found a match.

```
Chain CNI-ADMIN (1 references)
target     prot opt source               destination

Chain CNI-FORWARD (1 references)
target     prot opt source               destination
CNI-ADMIN  all  --  anywhere             anywhere             /* CNI firewall plugin rules */
ACCEPT     all  --  anywhere             10.80.0.2            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.2            anywhere
ACCEPT     all  --  anywhere             10.80.0.3            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.3            anywhere
ACCEPT     all  --  anywhere             10.80.0.4            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.4            anywhere
ACCEPT     all  --  anywhere             10.80.0.5            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.5            anywhere
ACCEPT     all  --  anywhere             10.80.0.6            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.6            anywhere
ACCEPT     all  --  anywhere             10.80.0.7            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.7            anywhere
ACCEPT     all  --  anywhere             10.80.0.8            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.8            anywhere
ACCEPT     all  --  anywhere             10.80.0.9            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.9            anywhere
ACCEPT     all  --  anywhere             10.80.0.10            ctstate RELATED,ESTABLISHED
ACCEPT     all  --  10.80.0.10            anywhere
```

Like you said this would happen for the vast majority of checks and would be reduced to a single check if we add the bypass for ESTABLISHED/RELATED connections. Will update soon, thanks!

Reply:
Oh, neat! Good find